src/rust/src/buf.rs

```diff
@@ -2,55 +2,97 @@
+        )
+    }
+}
 
-pub(crate) struct CffiBuf<'p> {
-    pyobj: pyo3::Bound<'p, pyo3::PyAny>,
-    _bufobj: pyo3::Bound<'p, pyo3::PyAny>,
-    buf: &'p [u8],
+#[cfg(Py_3_11)]
+fn _extract_buffer_length(
+    pyobj: &pyo3::Bound<'_, pyo3::PyAny>,
+    mutable: bool,
+) -> pyo3::PyResult<(pyo3::buffer::PyBuffer<u8>, *mut std::os::raw::c_void, usize)> {
+    let bufobj = pyo3::buffer::PyBuffer::<u8>::get(pyobj).map_err(|_| {
+        let errmsg = generate_non_convertible_buffer_error_msg(pyobj);
+        pyo3::exceptions::PyTypeError::new_err(errmsg)
+    })?;
+    if mutable && bufobj.readonly() {
+        return Err(pyo3::exceptions::PyTypeError::new_err(
+            "Buffer is not writable.",
+        ));
+    };
+    let buf_ptr = bufobj.buf_ptr();
+    let len = bufobj.len_bytes();
+    Ok((bufobj, buf_ptr, len))
 }
 
+#[cfg(not(Py_3_11))]
 fn _extract_buffer_length<'p>(
     pyobj: &pyo3::Bound<'p, pyo3::PyAny>,
     mutable: bool,
-) -> pyo3::PyResult<(pyo3::Bound<'p, pyo3::PyAny>, usize)> {
+) -> pyo3::PyResult<(pyo3::Bound<'p, pyo3::PyAny>, usize, usize)> {
     let py = pyobj.py();
     let bufobj = if mutable {
-        let kwargs = [(pyo3::intern!(py, "require_writable"), true)].into_py_dict(py)?;
+        let kwargs = pyo3::types::IntoPyDict::into_py_dict(
+            [(pyo3::intern!(py, "require_writable"), true)],
+            py,
+        )?;
         types::FFI_FROM_BUFFER
             .get(py)?
             .call((pyobj,), Some(&kwargs))
     } else {
         types::FFI_FROM_BUFFER.get(py)?.call1((pyobj,))
     }
     .map_err(|_| {
-        let errmsg = if pyobj.is_instance_of::<pyo3::types::PyString>() {
-            format!(
-                "Cannot convert \"{}\" instance to a buffer.\nDid you mean to pass a bytestring instead?",
-                pyobj.get_type()
-            )
-        } else {
-            format!("Cannot convert \"{}\" instance to a buffer.", pyobj.get_type())
-        };
+        let errmsg = generate_non_convertible_buffer_error_msg(pyobj);
         pyo3::exceptions::PyTypeError::new_err(errmsg)
     })?;
     let ptrval = types::FFI_CAST
         .get(py)?
         .call1((pyo3::intern!(py, "uintptr_t"), bufobj.clone()))?
         .call_method0(pyo3::intern!(py, "__int__"))?
         .extract::<usize>()?;
-    Ok((bufobj, ptrval))
+    let len = bufobj.len()?;
+    Ok((bufobj, ptrval, len))
+}
+
+pub(crate) struct CffiBuf<'p> {
+    pyobj: pyo3::Bound<'p, pyo3::PyAny>,
+    #[cfg(not(Py_3_11))]
+    _bufobj: pyo3::Bound<'p, pyo3::PyAny>,
+    #[cfg(Py_3_11)]
+    _bufobj: pyo3::buffer::PyBuffer<u8>,
+    buf: &'p [u8],
 }
 
 impl<'a> CffiBuf<'a> {
     pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
+        #[cfg(Py_3_11)]
+        let bufobj = {
+            let py_bytes = pyo3::types::PyBytes::new(py, buf);
+            pyo3::buffer::PyBuffer::get(&py_bytes)
+                .expect("Cannot convert \"bytes\" instance to a buffer.")
+        };
```

Comment:
This gives us extra allocations that defeat the point of the methods :-) My PoC just makes `bufobj` be `Option<Buffer>` on py311+ and stores a none there for this reason.

Reply:
Thanks for the feedback, and thank you for your patience! I overlooked the `Option` change in your diff while focusing on the `cfg` updates. I'll make the necessary adjustments and push an updated change shortly.

Reply:
I've implemented the changes in commit 1024811bb. Thanks again for your patience and understanding!

Reply:
Resolving per https://github.com/pyca/cryptography/pull/12857#issuecomment-2852370426